Update for the new Grid (#5)
DevExpress-Examples/blazor-default-settings: Readme.md

# DevExpress Blazor Components - Set predefined settings for the specific component

> Controls usually have the same settings for different pages of a large web application, and we recommend that you specify the component settings in markup. You can also override the **SetParametersAsyncCore** (for Grid) or **SetParametersAsync** (for other components) method to specify the default settings for Blazor components in your application. However, we do not recommend that you use this technique because it is uncommon in Blazor. Use it at your own risk.

Reviewer: ```suggestion
Controls usually have the same settings for different pages of a large web application. We recommend that you specify these component settings in markup. While you can also override the **SetParametersAsyncCore** (for the Grid) or the **SetParametersAsync** (for other components) method to specify default settings for Blazor components in your application, we do not recommend that you use this technique. ........

```
This note is from me and Vova: "uncommon" and "at your own risk" are vague and don't help usunderstand why one way is our recommendation and another isn't. And the article actually demonstrates this "uncommon", "do-it-at-your-own-risk" technique... thus, Blazor should allow other ways to create your own component with changed default settings, if we don't recommend that they override those methods. Can you please rewrite this part with specifics of why we do or don't.

Author: I moved the specifics part to the next paragraph.

```suggestion
Controls usually have the same settings for different pages of a large web application. We recommend that you specify these component settings in markup. You can also override the **SetParametersAsyncCore** (for the Grid) or the **SetParametersAsync** (for other components) method to specify the default settings for Blazor components in your application.

```

> Controls usually have the same settings for different pages of a large web application, and we recommend that you specify the component settings in markup. You can also override the **SetParametersAsyncCore** (for Grid) or **SetParametersAsync** (for other components) method to specify the default settings for Blazor components in your application. However, we do not recommend that you use this approach because it is not a common approach in Blazor. Use it at your own risk.

Reviewer 2: ```suggestion
Controls usually have the same settings for different pages of a large web application, and we recommend that you specify the component settings in markup. You can also override the **SetParametersAsyncCore** (for Grid) or **SetParametersAsync** (for other components) method to specify the default settings for Blazor components in your application. However, we do not recommend that you use this technique because it is uncommon in Blazor. Use it at your own risk.
```

![Grid with predefined settings and default](images/result.png)

> Create an inheritor from our Blazor control that overrides the **SetParametersAsync** or **SetParametersAsyncCore** method. In this method, set the default values for the public properties before you call the base method. You can initialize properties with custom default values in markup because the base method is called after custom settings are applied. We do not recommend that you override and change the values of protected properties.

Reviewer 2: > Create an inheritor from our Blazor control that overrides the **SetParametersAsync** or **SetParametersAsyncCore** method.

I see a logical flaw in this sentence. You (or, rather, the original author) clearly meant that "an inheritor" overrides the mentioned methods. However, it reads "our Blazor control that overrides the **SetParameterAsync** or **SetParameterAsyncCore** method."

> In this method, set the default values for the public properties before you call **the base method**.

I'm unsure what you meant by "the base method". Do you speak of the corresponding (overridden) method in the base class (that is, your Blazor component class).

> You can initialize properties with **custom default values**

I have only one question. How default values can be custom or vice versa? It sounds like an oxymoron to me.

> We do not recommend that you override and change the values of protected properties.

We could have clarified what would happen otherwise, that is, a rationale for this kind of limitation.

Author: > We could have clarified what would happen otherwise, that is, a rationale for this kind of limitation.

In this case, we cannot guarantee the correct functioning of our components. That is why these properties are not public but protected.

```suggestion
Inherit from our Blazor control to override the **SetParametersAsync** or **SetParametersAsyncCore** method. In this method, set the default values for the public properties before you call the base **SetParametersXXX** method. In markup, you can initialize properties with other than pre-defined default values because the base method is called after custom settings are applied. We do not recommend that you override and change the values of protected properties.

```

Reviewer 2: I often say that one shouldn't rely on non-public APIs mainly because they are subject to change in future builds. We can only guarantee backward compatibility for the "white" or "official" APIs.

Regarding your change:

```suggestion
Inherit from our Blazor control to override the **SetParametersAsync** or **SetParametersAsyncCore** method. In one of these methods, set the default values for the public properties before you call the base **SetParametersXXX** method. In markup, you can initialize properties with modified default values because the base method is called after custom settings are applied. We do not recommend that you override and change the values of protected properties.
```

> Inherit from our Blazor control to override the **SetParametersAsync** or **SetParametersAsyncCore** method. In one of these methods, set the default values for the public properties before you call the base **SetParametersXXX** method. In markup, you can initialize properties with modified default values because the base method is called after custom settings are applied. We do not recommend that you override and change the values of protected properties.

Reviewer: ```suggestion
Inherit [inherit what?] from our Blazor control to override the **SetParametersAsync** or **SetParametersAsyncCore** method. In one of these methods [which one? we should be specific - see rephrase below comment], set the default values for the public properties before you call the base **SetParametersXXX** method. In markup, you can initialize properties with modified default values since the base method is called after custom settings are applied. We do not recommend that you override and change the values of protected properties. [ can we tell them why we don't recommend overriding/changing the values of protected properties?]

```
Rewrite. Depending on what is technically right. Since I don't know, I am providing several rewrites.
 You can set the default values for the public properties in either the **SetParametersAsync** or **SetParametersAsyncCore** method before you call the base **SetParametersXXX** method.

Is there a different outcome if they set default values for the public properties in **SetParametersAsync** vs. **SetParametersAsyncCore**?

Author: About protected properties: We do not document protected properties and do not guarantee that we don't change anything in future updates. Also, if someone decides to change anything in our protected members, they should be 100% sure what exactly they are trying to do. Otherwise, the control would behave incorrectly.
SetParametersAsync vs. SetParametersAsyncCore: the choice depends on the internal implementation of our components. I mention in the previous paragraph what method to use with what component.

```suggestion
Create an inheritor from our Blazor control to override the **SetParametersAsync** or **SetParametersAsyncCore** method. You can set the default initial values for the **public** properties in either the SetParametersAsync or SetParametersAsyncCore method before you call the base SetParametersXXX method. In markup, you can initialize properties with modified default values since the base method is called after custom settings are applied. 

Note that we do not recommend that you override and change the values of **protected** properties as it may lead to incorrect control behavior. Do it at your own risk.
```

Reviewer: Note that we do not recommend that you override and change the values of **protected** properties as it may lead to incorrect control behavior.

Author: ```suggestion
Create an inheritor from our Blazor control to override the **SetParametersAsync** or **SetParametersAsyncCore** method. You can set the default initial values for the **public** properties in either the SetParametersAsync or SetParametersAsyncCore method before you call the base SetParametersXXX method. In markup, you can initialize properties with modified default values since the base method is called after custom settings are applied. 

Note that we do not recommend that you override and change the values of **protected** properties as it may lead to incorrect control behavior.
```

> In this example, the [MyGrid](./CS/DxBlazorComponentsDefaultSettings/Components/MyGrid.cs) component contains the inherited **DxGrid** component whose **PageSize**, **ShowFilterRow**, **AllowSort**, and **ShowGroupPanel** properties contain new default values. You can compare the **MyGrid** component to the original **DxGrid** component on the [Index](./CS/DxBlazorComponentsDefaultSettings/Pages/Index.razor) page to ensure they have the same settings in markup.

Reviewer: ```suggestion
In this example [which example? there was another paragraph before this, not an example; are you referring to the image? If so, please clarify.], the [MyGrid](./CS/DxBlazorComponentsDefaultSettings/Components/MyGrid.cs) component contains the inherited **DxGrid** component whose **PageSize**, **ShowFilterRow**, **AllowSort**, and **ShowGroupPanel** properties contain new default values. You can compare the **MyGrid** component to the original **DxGrid** component on the [Index](./CS/DxBlazorComponentsDefaultSettings/Pages/Index.razor) page to ensure they have the same settings in markup.

```

Author: I refer here to source code files.

```suggestion
In this example solution, the [MyGrid](./CS/DxBlazorComponentsDefaultSettings/Components/MyGrid.cs) component contains the inherited **DxGrid** component whose **PageSize**, **ShowFilterRow**, **AllowSort**, and **ShowGroupPanel** properties contain new default values. You can compare the **MyGrid** component to the original **DxGrid** component on the [Index](./CS/DxBlazorComponentsDefaultSettings/Pages/Index.razor) page to ensure they have the same settings in markup.

```

> In this example, the [MyGrid](./CS/DxBlazorComponentsDefaultSettings/Components/MyGrid.cs) component contains the inherited **DxGrid** component whose **PageSize**, **ShowFilterRow**, **AllowSort**, and **ShowGroupPanel** properties contain new default values. You can compare the **MyGrid** component with the default **DxGrid** component on the [Index](./CS/DxBlazorComponentsDefaultSettings/Pages/Index.razor) page to ensure they have the same settings in markup.

Reviewer 2: ```suggestion
In this example, the [MyGrid](./CS/DxBlazorComponentsDefaultSettings/Components/MyGrid.cs) component contains the inherited **DxGrid** component whose **PageSize**, **ShowFilterRow**, **AllowSort**, and **ShowGroupPanel** properties contain new default values. You can compare the **MyGrid** component to the original **DxGrid** component on the [Index](./CS/DxBlazorComponentsDefaultSettings/Pages/Index.razor) page to ensure they have the same settings in markup.
```

> You cannot use this approach for **DxChartXXXSeries** components and specify the **ValueField**, **ArgumentField**, **NameField**, and **PaneField** properties.

Reviewer 2: Ray discouraged usage of the word "approach". That's why I always change it.

```suggestion
You cannot use this technique for **DxChartXXXSeries** components and specify the **ValueField**, **ArgumentField**, **NameField**, and **PaneField** properties.
```

> You cannot use this technique for **DxChartXXXSeries** components and specify the **ValueField**, **ArgumentField**, **NameField**, and **PaneField** properties.

Reviewer: ```suggestion
You cannot use this technique for **DxChartXXXSeries** components and specify the **ValueField**, **ArgumentField**, **NameField**, and **PaneField** properties.

```
You haven't mentioned **DxChartXXXSeries** components before. How does this relate to this help doc? This last paragraph doesn't seem to flow with the one before. Is there something missing here that should be said. Let me know.

Author: As the example title says, the approach shown here can be applied to all our Blazor components. However, the internal implementation of the Chart series requires specifying **ValueField**, **ArgumentField**, **NameField**, and **PaneField** directly in markup to define series data types and values correctly. So, this is a limitation of the approach demonstrated here.

Reviewer: ok
